File: day1/code.py
import os

def solve1():
    with open('input.txt') as finput:
        sum = 0
        for s in finput:
            x, posx, y, posy = 0, len(s), 0, -1
            for i in range(1, 10):
                p = s.find(str(i))
                if p != -1 and p < posx:
                    x, posx = i, p 
                p = s.rfind(str(i))
                if p != -1 and p > posy:
                    y, posy = i, p
            sum += x * 10 + y
        print("task1: ", sum)
        
def solve2():
    with open('input.txt') as finput:
        sum = 0
        m = {
            "one"   : 1,
            "1"     : 1,
            "two"   : 2,
            "2"     : 2,
            "three" : 3, 
            "3"     : 3,
            "four"  : 4,
            "4"     : 4,
            "five"  : 5,
            "5"     : 5,
            "six"   : 6,
            "6"     : 6,
            "seven" : 7,
            "7"     : 7,
            "eight" : 8,
            "8"     : 8,
            "nine"  : 9,
            "9"     : 9,
        }
        for s in finput:
            x, posx, y, posy = 0, len(s), 0, -1
            for ss in m: 
                p = s.find(ss)
                if p != -1 and p < posx:
                    x, posx = m[ss], p 
                p = s.rfind(ss)
                if p != -1 and p > posy:
                    y, posy = m[ss], p
            sum += x * 10 + y
        print("task2: ", sum)
        
solve1()
solve2()

File: day1/code.cpp
#include <bits/stdc++.h>
using namespace std;

void task1(vector<string> str) {
    int sum = 0;
    for (string s : str) {
        int x = 0, y = 0;
        for (auto ch : s)
            if (isdigit(ch)) {
                x = ch - '0';
                break;
            }
        reverse(s.begin(), s.end());
        for (auto ch : s)
            if (isdigit(ch)) {
                y = ch - '0';
                break;
            }
        sum += x * 10 + y;
    }
    cout << "task1: " << sum << endl;
}

void task2(vector<string> str) {
    map<string, int> m;
    m["1"] = m["one"] = 1;
    m["2"] = m["two"] = 2;
    m["3"] = m["three"] = 3;
    m["4"] = m["four"] = 4;
    m["5"] = m["five"] = 5;
    m["6"] = m["six"] = 6;
    m["7"] = m["seven"] = 7;
    m["8"] = m["eight"] = 8;
    m["9"] = m["nine"] = 9;
    int sum = 0;
    for (string s : str) {
        int x = 0, y = 0;
        for (int i = 0; i < s.length(); ++i) {
            bool f = false;
            for (auto [ss, v] : m)
                if (ss.size() + i - 1 < s.length()) {
                    if (ss == s.substr(i, ss.size())) {
                        x = v, f = true;
                        break;
                    }
                }
            if (f)
                break;
        }
        for (int i = s.length() - 1; i >= 0; --i) {
            bool f = false;
            for (auto [ss, v] : m)
                if (ss.size() + i - 1 < s.length())
                    if (ss == s.substr(i, ss.size())) {
                        y = v, f = true;
                        break;
                    }
            if (f)
                break;
        }
        sum += x * 10 + y;
    }
    cout << "task2: " << sum << endl;
}

int main () {
    freopen("input.txt", "r", stdin);
    vector<string> str;
    string s;
    while (getline(cin, s))
        str.push_back(s);
    task1(str);
    task2(str);
    return 0;
}